### PR TITLE
Add `:relative-time-interval` mbql function

### LINKED ISSUE #34573: Relative dates filter uses invalid date range
### Describe the bug

When using "relative dates" filter option, invalid range is returned due to incorrect end date.

### To Reproduce

1. Use date range filter and select "Relative dates..." 
2. Enter: "Past 1 month"
3. Scroll down to '...' and add "Starting from..."
4. Enter: "Starting from 2 months ago"
5. Click Update filter

Screenshot 2023-10-11 at 17 24 45

Even though on the screenshot the selected date range is shown as "Jul 1 - Jul 31", database query actually is between "Jul 1 00:00 - Aug 1 00:00".

In my case, that date field is truncated to YYYY-MM only and that causes the query to encompass all Jul and Aug.

### Expected behavior

Filter should only query "Jul 1 00:00:00 - Jul 31 23:59:59"

### Logs

This filter adds the following lines to the query:

```sql
SELECT (...) AS "source"
WHERE
  DATE_ADD('month', 2,
    DATE_TRUNC('month',"source"."billing_period")
  ) 
  BETWEEN DATE_TRUNC('month', DATE_ADD('month', -1, NOW())) AND DATE_TRUNC('month', NOW())
```

We can run a simple query against Athena to check those values and confirm that the end date is incorrect:
```sql
SELECT 
  DATE_TRUNC('month', DATE_ADD('month', -1, NOW())) AS date_from,
  DATE_TRUNC('month', NOW()) AS date_to

--Result
-- date_from: 2023-9-1, 12:00 AM
-- date_to: 2023-10-1, 12:00 AM
```

### Information about your Metabase installation

```JSON
- Queried database: AWS Athena
- Internal database: Postgres
- Metabase version: 0.73.3
- OS: Ubuntu
```

### Severity

medium

### Additional context

_No response_

### LINKED ISSUE #45942: Filtering on date columns by relative dates using weeks results in wrong time range
### Describe the bug

This is a follow-up bug for https://github.com/metabase/metabase/issues/42291. I decided to create a new GH issue as opposed to re-opening since it's a slightly different issue this time.

We are still seeing duplicate results in our question, week over week, around the beginning and end of the time range.
Context: we have an alert setup every week for a question that uses a filter on created_at in the previous week, 16 weeks ago. 

This seems to because the time range that is generated is 8 days instead of 7, apparently.

### To Reproduce

1. Go to https://stats.metabase.com/question/4859
2. Click on Show filters
3. Click on Created At is in the previous week, 16 weeks ago
4. There is a preview in the UI for the time range: March 24-30 (see screenshot)
image

5. Notice there is one result with Created At on March 24
6. Now change the filter to 17 weeks ago, the preview in the UI states shows March 17 and 23
image

7. Notice the same record in the result set with Created on March 24, which should not be included

### Expected behavior

- The time range applied in the where clause matches the time range I see in the UI when I select a value
- The time range is 7 days, e.g. from March 17 00:00 to March 23 23:59, no results from March 24
- There is no overlap as you increment or decrement the offset in weeks, otherwise you end up with duplicates

### Logs

_No response_

### Information about your Metabase installation

```JSON
- Reproduced on stats, Built on 2024-07-22, Hash: e02589a
```

### Severity

P1

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by lbrdnk**
Closes https://github.com/metabase/metabase/issues/45942
Closes https://github.com/metabase/metabase/issues/34573

### Description

This PR adds `:relative-time-interval` mbql function and modifies FE to generate relative datetime filters using it.

Function has a following signature `column, value, bucket, offsetValue, offsetBucket`. `options` are not used as I found them redundant at this point. `value` and `offsetValue` are allowed to be integer only at this point.

Function is desugared into form as follows `[:and [:>=..] [:<..]]` in desugar qp middleware.

More context [on slack](https://metaboat.slack.com/archives/C04DN5VRQM6/p1721849734133429).

### How to verify

Run modified and new tests, or verify manually following reproductions in the linked issues.

### Checklist

- [X] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

File: src/metabase/legacy_mbql/util.cljc [hunk 1]
```diff
@@ -234,6 +234,27 @@
...
+                       [:relative-datetime 1 bucket])
+         upper-bound (if (neg? value)
+                       [:relative-datetime 0 bucket]
+                       [:relative-datetime (inc value) bucket])
+         lower-with-offset [:+ lower-bound offset]
+         upper-with-offset [:+ upper-bound offset]]
+     [:and
+      [:>= col-default-bucket lower-with-offset]
+      [:<  col-default-bucket upper-with-offset]])))
+
```
Comment (lbrdnk) [inline src/metabase/legacy_mbql/util.cljc:257]: Logic that generates intervals now live here. Change also fixes positive shift that worked incorrectly before.

**Review by ranquild — COMMENTED** [review]
@lbrdnk `options` are used for `include-current` flag. I see we don't have e2e tests for this case which is bad. But apart from that your change looks awesome!
Screenshot 2024-07-30 at 11 01 20

**Comment by lbrdnk** [comment]
@ranquild, my understanding is, that if the _offset_ relative date filter is in use, we provide no current option. I think adding options to `relative-time-interval` will make sense in a follow up, _if we decide_ to remove the `time-interval` in favor of `relative-time-interval`.

That would require further modifications to `relative-time-interval`, namely aforementioned _include current_ and update to handling of empty offset (I've set the schema to int atm).

That would be ihmo step the right direction because we could shrink the code base (even just a small) bit and make things less complex.

**Comment by lbrdnk** [comment]
That said, if you expand up the change in `frontend/src/metabase-lib/filter.ts`, you'll see the good old `time-interval` is still in use there for the _current_ case. As per my previous comment, I believe removing it (or changing `time-interval` to use `relative-time-interval` logic) could make a good follow up to this PR.

**Review by ranquild — APPROVED** [review]
@lbrdnk got it, thanks for explaining

**Review by bshepherdson — APPROVED** [review]

**Pull request merged**